lib/DragNDrop.pm

```diff
@@ -0,0 +1,228 @@
+at the bottom of the bucket.
+(The first created bucket may never be removed.)
+###############################################################################
+
+To output the bucket pool to HTML, call:
+
+ $bucket_pool->HTML
+
+To output the bucket pool to LaTeX, call:
+
+ $bucket_pool->TeX
+
+=head1 EXAMPLES
+
+See draggableProof.pl and draggableSubsets.pl
+
+=cut
+
+###############################################################################
+
+use strict;
+use warnings;
+
+package DragNDrop;
+
+sub new {
+	my $self = shift;
+	my $class = ref($self) || $self;
+
+	# 'id' of html <input> tag corresponding to the answer blank. Must be unique to each pool of DragNDrop buckets
+	my $answerInputId = shift;
+
+	# array of all statements provided
+	my $aggregateList = shift;
+
+	# instructor-provided default buckets with pre-included statements encoded
+	# by the array of corresponding statement indices
+	my $defaultBuckets = shift;
+
+	my %options = (
+	AllowNewBuckets => 0,
+	@_
+	);
+
+	$self = bless {
+		answerInputId => $answerInputId,
+		bucketList => [],
+		aggregateList => $aggregateList,
+		defaultBuckets => $defaultBuckets,
+		%options,
+	}, $class;
+
+	return $self;
+}
+
+sub addBucket {
+	my $self = shift;
+
+	my $indices = shift;
+
+	my %options = (
+	label => "",
+	removable => 0,
+	@_
+	);
+
+	my $bucket = {
+		indices => $indices,
+		list => [ map { $self->{aggregateList}->[$_] } @$indices ],
+		bucket_id => scalar @{ $self->{bucketList} },
+		label => $options{label},
+		removable => $options{removable},
+	};
+	push(@{$self->{bucketList}}, $bucket);
+
+}
+
+sub HTML {
+	my $self = shift;
+
+	my $out = '';
+	$out .= "<div class='bucket_pool' data-ans='$self->{answerInputId}'>";
+
+	# buckets from instructor-defined default settings
+	for (my $i = 0; $i < @{$self->{defaultBuckets}}; $i++) {
+		my $defaultBucket = $self->{defaultBuckets}->[$i];
+		$out .= "<div class='hidden default bucket' data-bucket-id='$i' data-removable='$defaultBucket->{removable}'>";
+		$out .= "<div class='label'>$defaultBucket->{label}</div>";
+		$out .= "<ol class='answer'>";
+		for my $j ( @{$defaultBucket->{indices}} ) {
+			$out .= "<li data-shuffled-index='$j'>$self->{aggregateList}->[$j]</li>";
+		}
+		$out .= "</ol></div>";
+	}
+
+	# buckets from past answers
+	for my $bucket ( @{$self->{bucketList}} ) {
+		$out .= "<div class='hidden past_answers bucket' data-bucket-id='$bucket->{bucket_id}' ";
+		$out .= "data-removable='$bucket->{removable}'>";
+		$out .= "<div class='label'>$bucket->{label}</div>";
+		$out .= "<ol class='answer'>";
+
+		for my $index ( @{$bucket->{indices}} ) {
+			$out .= "<li data-shuffled-index='$index'>$self->{aggregateList}->[$index]</li>";
+		}
+		$out .= "</ol>";
+		$out .= "</div>";
+	}
+	$out .= '</div>';
+	$out .= "<br clear='all'><div><a class='btn reset_buckets'>reset</a>";
+	if ($self->{AllowNewBuckets} == 1) {
+		$out .= "<a class='btn add_bucket' data-ans='$self->{answerInputId}'>add bucket</a>";
```

Comment:
It would be good to add the css `btn-secondary` class to both of these buttons.  This class does not exist in bootstrap version 2, and so at this point it will do nothing.  However, soon we will be upgrading to bootstrap version 5, and with version 5 the `btn` class is not intended to be used by itself.  The `btn-secondary` class is the flavor that will work best here.  If you don't do this now, I will add it later.  I have a pull request coming soon that implements this upgrade.

Reply:
Thank you for the comments! I have updated the files per your suggestions. I will update the linked PR on openwebwork/webwork2 next.